Add option to convert cooling times in output file to seconds

anu1217 opened the pull request:
This PR introduces changes that allow the user to select the units (default or seconds) in which cooling times are printed in the headers of the output tables, if the `units` keyword is added in the `output` section of the input file.

Adding this in as a draft for now as this change has had many moving pieces, with changes to function definitions in multiple files, and I anticipate more work being necessary for everything to work as intended.

fixes #137

anu1217 committed:
add cooling time variables

anu1217 committed:
add cooling time integers

anu1217 committed:
add Output_def.h to CoolingTime.h

anu1217 committed:
add cooltime_units as argument to writeTotalHeader()

anu1217 committed:
add condition to convert units for total header

anu1217 committed:
add cooltimeUnits

anu1217 committed:
add cooltimeType as argument to setNorm()

anu1217 committed:
add cooltime_units to Result

anu1217 committed:
add missing ;

anu1217 committed:
turn getCooltimeMode into a static function

anu1217 committed:
change name of returned object

anu1217 committed:
reflect getCooltimeMode being changed to static

anu1217 committed:
only declare function in Result.h

anu1217 committed:
add logic to writeHeader()

anu1217 committed:
fix arguments for writeHeader()

anu1217 committed:
fix logic

gonuke commented on src/Loading.C:
```diff
@@ -318,7 +320,7 @@ void Loading::tally(Result *volOutputList, double vol)
     output information can be given. */
 void Loading::write(int response, int writeComp, CoolingTime* coolList, 
-		    int targetKza, int normType)
+		    int targetKza, int normType, const OutputFormat* outFmt)
```
Why do you need `outFmt` here?  it doesn't appear to be used

And if you don't need it, then you don't need to include it's header above.

gonuke commented on src/CoolingTime.C:
```diff
@@ -131,61 +132,92 @@ int CoolingTime::makeCoolingTimes(double *& coolingTimes)
 }
 
-void CoolingTime::getCoolTimesStrings(std::vector<std::string>& coolTimesList)
+void CoolingTime::getCoolTimesStrings(std::vector<std::string>& coolTimesList, const OutputFormat* outFmt)
```
The header defines this as an integer not a a pointer to type `OutputFormat`.  I think I like that choice better since it means you don't need to test if `outFmt` is defined below.

gonuke commented on src/OutputFormat.C:
I think we need to write the cooling time units for these ones, too, probably

gonuke commented on src/Volume.C:
```diff
@@ -617,7 +619,7 @@ void Volume::postProc()
     can be given. */
 void Volume::write(int response, int writeComp, CoolingTime* coolList, 
-		   int targetKza, int normType)
+		   int targetKza, int normType, const OutputFormat* outFmt)
```
Why do you need `outFmt` here?  it doesn't appear to be used.

And if you don't need it, then you don't need to include it's header above.

gonuke commented on src/Result.C:
```diff
@@ -607,6 +614,16 @@ void Result::setNorm(double passedActMult, int normType)
+    break;
+  case COOLTIME_DEF:
+    cooltime_units = 1;
+    break;
```
Don't need these lines since the same thing happens in both cases

```suggestion
```

gonuke commented on src/Result.C:
```diff
@@ -607,6 +614,16 @@ void Result::setNorm(double passedActMult, int normType)
+    break;
+  default:
+    cooltime_units = 1;
+  }
```
Why not just:

```suggestion
  cooltime_units = cooltimeType;
```

gonuke reviewed (changes requested):
I'm a little surprised that this compiles, given some mismatches in the header and code.

anu1217 committed:
Apply suggestion from @gonuke

Co-authored-by: Paul Wilson <paul.wilson@wisc.edu>

anu1217 committed:
Apply suggestion from @gonuke

Co-authored-by: Paul Wilson <paul.wilson@wisc.edu>

anu1217 committed:
Apply suggestion from @gonuke

Co-authored-by: Paul Wilson <paul.wilson@wisc.edu>

anu1217 committed:
Merge branch 'main' into output_s

anu1217 committed:
Merge branch 'svalinn:main' into output_s

anu1217 committed:
Merge branch 'output_s' of github.com:anu1217/ALARA into output_s

anu1217 committed:
add missing curly braces

anu1217 committed:
add condition to ensure cooltime units are printed to top header

anu1217 committed:
redefine char

anu1217 committed:
remove conditional statement

anu1217 committed:
remove commented lines

anu1217 commented:
Before marking this as ready for review, I changed my approach to this problem and pushed all my changes to a different branch, before merging the new branch into this one. I believe this didn't execute the way I expected and left some extraneous changes, which I think have all been resolved now. The folded dose/WDR still need to be modified to print out the cooling time in the desired units.

anu1217 committed:
add cooltimeUnits to other responses

gonuke commented on src/OutputFormat.C:
```diff
@@ -175,7 +183,25 @@ OutputFormat* OutputFormat::getOutFmts(istream& input)
+	  
+	  delete[] next->cooltimeUnits;
+	  input >> token;
+	  next->cooltimeUnits = new char[strlen(token)+2];
```
For better or worse, this is not backwards compatible and will make old input files fail. We should discuss how important that is...

gonuke reviewed (changes requested):
Sorry - this review was 1/2 done a while ago and got lost...

anu1217 committed:
remove unused header

correctly set metric multiplier and cooltime units

remove unused break

address feedback

anu1217 committed:
ensure backwards compatibility if cooltimeUnits are not added

add std namespace

avoid crossing initialization

move initialization of pos

declare pos in block

remove extra space

reinitialize cooltimeUnits so that there is no extra space

initialize cooltimeUnits with the length of the units token

account for nonexisting cooling time units

check condition with first character of token

anu1217 committed:
Merge pull request #6 from anu1217/test_output_s

Address feedback and test some changes

anu1217 reviewed (commented):
I've addressed most of the feedback here and have tested these changes in a container.

gonuke commented on src/OutputFormat.C:
```diff
@@ -141,6 +149,7 @@ OutputFormat* OutputFormat::getOutFmts(istream& input)
       switch (1<<type)
 	{
 	case OUTFMT_UNITS:
+	  {	
```
Do we need these braces?

anu1217 replied:
We need these braces due to the initialization of `std::streampos pos` in the case, or we get a "crosses initialization" error during compilation

gonuke replied:
So if we end up getting rid of the `streampos` (which my gut tells me we probably will) then we can get rid of these, too.

gonuke commented on src/OutputFormat.C:
```diff
@@ -175,7 +184,34 @@ OutputFormat* OutputFormat::getOutFmts(istream& input)
-	  break;
+	  
+	  delete[] next->cooltimeUnits;
+	  next->cooltimeUnits = new char[strlen(token)+1];
```
Perhaps we figure out a way to make it optional?

anu1217 replied:
It should be optional now since any token (or lack thereof) that doesn't start with "s" is taken to be the default case

gonuke commented on src/OutputFormat.C:
```diff
@@ -175,7 +184,34 @@ OutputFormat* OutputFormat::getOutFmts(istream& input)
+			{
+			error(230, "Unknown cooling time unit '%s'", token);		
+			}
+		}
```
Why might need to implement clang-formatting next, but this style is preferred:

```suggestion
	  if (input >> token) {
		if (tolower(token[0]) == 's') {
			next->cooltimeType = COOLTIME_S;

			delete[] next->cooltimeUnits;
			next->cooltimeUnits = new char[strlen(token)+1];
			strcpy(next->cooltimeUnits, token);
		} else if (tolower(token[0]) == 'd') {
			input.seekg(pos);
		} else {
			error(230, "Unknown cooling time unit '%s'", token);		
		}
	}
```

gonuke commented on src/Result.C:
```diff
@@ -592,12 +593,19 @@ void Result::readDump()
 
-  switch (normType) {
+  switch (normType) 
+  {
```
```suggestion
  switch (normType) {
```

gonuke reviewed (changes requested):
Sorry for the delay - I had some pending changes that I forgot to submit.

anu1217 committed:
Merge branch 'svalinn:main' into output_s

anu1217 committed:
Merge branch 'svalinn:main' into output_s

anu1217 committed:
change formatting

Co-authored-by: Paul Wilson <paul.wilson@wisc.edu>

anu1217 committed:
change formatting

Co-authored-by: Paul Wilson <paul.wilson@wisc.edu>

anu1217 committed:
change logic and formatting

anu1217 commented on src/OutputFormat.C:
```diff
@@ -175,7 +184,34 @@ OutputFormat* OutputFormat::getOutFmts(istream& input)
+			}
+		else if (tolower(token[0]) == 'd')
+			{
+			input.seekg(pos);
```
I've changed the logic used here, but since `while (strcmp(token,"end"))` is reading the list of output types until `end` has been found, doing nothing in this case would consume the `end` token earlier than we want

gonuke replied:
Right - but only if we try to make this new addition optional on this same `units` line.

gonuke commented on src/OutputFormat.C:
```diff
@@ -175,7 +184,28 @@ OutputFormat* OutputFormat::getOutFmts(istream& input)
+
+	  delete[] next->cooltimeUnits;
+	  next->cooltimeUnits = nullptr;
+	  std::streampos pos = input.tellg();
+	  
+	  if (input >> token) {
+		if (tolower(token[0]) == 's') {
+			next->cooltimeType = COOLTIME_S;
+
+			next->cooltimeUnits = new char[strlen(token)+2];
+			strcpy(next->cooltimeUnits, token);
+		} else {
+			input.seekg(pos);
+
+			next->cooltimeType = COOLTIME_DEF;
+			next->cooltimeUnits = new char[4];
+			strcpy(next->cooltimeUnits, "def");
+		}
+	}	
+	break;	
+	  }
```
This approach is fragile. For example, it fails if a user does NOT include the optional cooling time unit, and the next entry in the output format block is `specific_activity` - something that might be quite common.

gonuke replied:
Given the way that ALARA reads input files, I don't think we can avoid breaking backwards compatibility with this.

The only way might be to add a new optional line in the output format block just for this purpose.

gonuke reviewed (changes requested):
Seems we need to discuss whether this can be optional in this way.
